### PR TITLE
"Member functions" and "constructors".

### PR DESCRIPTION
**Pull request by dkorolev**
Hi @mzhurovich ,

Work in progress, just to make sure you're unblocked.

Thanks!
Dima

### CODE REVIEW

**Comment by sompylasar** [comment]
@dkorolev all the macros should be namespaced with a prefix (CURRENT_CONSTRUCTOR, CURRENT_RETURNS).

@mzhurovich I believe CURRENT_CONSTRUCTOR(int z) {} is better than CURRENT_CONSTRUCTOR(Serializable) (int z) {} because the wrapping typename "Serializable" is written once per struct declaration.

**Comment by mzhurovich** [comment]
@sompylasar +1 for prefixes.
As for constructors, it's now a trade off between declaration and usage - look at the code invoking constructor.

**Comment by sompylasar** [comment]
@mzhurovich yes, I see. The invocation and field initialization is
definitely prettier with the constructor being a real constructor.

**Comment by dkorolev** [comment]
Thanks @sompylasar @mzhurovich.

A natively invoked constructor is cleaner, let's take that route. I was uncertain whether one can `std::enable_if<>` a constructor, and, since we can, it makes sense.

Addressing other comments now.

Thanks,
Dima

**Comment by dkorolev** [comment]
[ Amended https://github.com/c5t/Current/pull/255, as this https://github.com/c5t/Current/pull/256 contains it. ]

@mzhurovich : Your suggestion with constructor syntax, thanks!
@sompylasar : Your suggestion with JSON error messages phrasing, thanks!

PTAL,
Dima

File: Reflection/struct.h [hunk 1]
```diff
@@ -107,6 +136,10 @@ struct CurrentStructFieldsConsistency<T, -1> {
     constexpr static const char* name() { return #s; }                                      \
     constexpr static size_t index_base = __COUNTER__;                                       \
     typedef CURRENT_STRUCT_IMPL_##s<::current::reflection::CountFields> field_count_struct; \
+    template <typename... ARGS>                                                             \
+    static ConvertibleInto<s, std::tuple<ARGS...>> Construct(ARGS&&... args) {              \
```
Comment (mzhurovich) [inline Reflection/struct.h]: Do we still need it?

Comment (dkorolev) [inline Reflection/struct.h]: :+1: removed in 07ab93b

File: Reflection/struct.h [hunk 2]
```diff
@@ -95,6 +95,21 @@ struct CurrentStructFieldsConsistency<T, -1> {
 #define CURRENT_STRUCT(...) \
   CURRENT_STRUCT_SWITCH(__VA_ARGS__, CURRENT_STRUCT_DERIVED, CURRENT_STRUCT_NOT_DERIVED)(__VA_ARGS__)
 
+template <typename T, typename IS>
+struct DoConstruct {};
+
+template <typename T, int... IS>
+struct DoConstruct<T, bricks::variadic_indexes::indexes<IS...>> {
```
Comment (mzhurovich) [inline Reflection/struct.h]: And this?

Comment (dkorolev) [inline Reflection/struct.h]: :+1: ich bin ein Idiot. f8289cf

**Comment by mzhurovich** [comment]
@dkorolev What do you think about prefixing `RETURNS` macro as well?

**Comment by dkorolev** [comment]
^ e1167d4 :-)

**Comment by mzhurovich** [comment]
@dkorolev Yay!
Ready to merge :)

**Comment by dkorolev** [comment]
@mzhurovich What a day! What can't be accomplished on such a splendid day!

(Slack also suggest I get enough sleep if I can.)

**Comment by mzhurovich** [comment]
@dkorolev I second Slack :)

**Pull request merged**